File: Day4/tasks/catkin_ws/src/package_2/src/publisher.cpp
#include "ros/ros.h"
#include "std_msgs/Int32.h"

//Full documentation: http://wiki.ros.org/ROS/Tutorials/WritingPublisherSubscriber%28c%2B%2B%29 


int main(int argc, char **argv)
{
	//Initize new ROS node named "talker"
	ros::init(argc, argv, "talker_node");

	//creating a node handle: it is reference assigned to a new node
	ros::NodeHandle nh;

	//create a publisher with a topic "count_topic" that will send a String msg
	// 1000 is the buffer size
	ros::Publisher chatter_publisher = nh.advertise<std_msgs::Int32>("count_topic", 1000);     //the publisher topic should be exactly the same as the subscription topic


	//Rate is a class used to define frequency for a loop.
	ros::Rate loop_rate(1); //msg each 1 sec

	int count = 0;

	while (ros::ok()) //keep spinning loop until user terminate
	{
		//create a new String ROS message:

		std_msgs::Int32 msg; //create a int for the data

		//assign the int data to ROS message data field
		msg.data = count;

		//Print the content of the message in the terminal 
		ROS_INFO("Counter: %d\n", msg.data);

		//Publish the message
		chatter_publisher.publish(msg);

		ros::spinOnce(); //Need to call this function often to allow ROS to process incoming messages
		
		loop_rate.sleep(); //Sleep for the rest of the cycle, to enforce the loop rate
		
		count ++;
	}

	return 0;
}

File: Day4/tasks/catkin_ws/src/package_2/scripts/publisher.py
#!/usr/bin/env python

#Full documentation: http://wiki.ros.org/ROS/Tutorials/WritingPublisherSubscriber%28python%29

import rospy #importing ros library in python
from std_msgs.msg import Int32 #it can be int or float


def talker():
    #creating the node:
    rospy.init_node("pub_counter_node", anonymous=True) 
    #In ROS, nodes are named uniqely. If 2 nodes with the same name are lunched, the previous one is kicked off.
    #ananymos=True makes rospy choose a uniqe name for each node so that multiple nodes can run simultaneously. 

    pub = rospy.Publisher("/countTopic", Int32, queue_size=10) 
    #the publisher topic should be exactly the same as the subscription topic
    #pub is a variable, (name of the topic, .., bufer: stadered 10 but it can be 30)

    rate = rospy.Rate(1)
    #hz: every 1 sec. a message will be sent

    count = 0 
    #this loop is terminated with ctrl C
    while not rospy.is_shutdown(): 
        rospy.loginfo(count) #to print
        pub.publish(count) #to publish
        rate.sleep() #this makes the loop works until ctrl C
        count = count + 1

#starting main function:    
if __name__ == '__main__':
    try:
        talker()
    except rospy.ROSInterruptException:
        pass

#OR: write the function directrly

    